client/src/main/java/io/lionweb/client/inmemory/RepositoryData.java

```diff
@@ -139,8 +172,115 @@ void retrieve(String nodeId, int limit, List<SerializedClassifierInstance> retri
+  public @NotNull ValidationResult checkConsistency() {
+    ValidationResult result = new ValidationResult();
+
+    // Check for invalid node IDs
+    for (String nodeId : nodesByID.keySet()) {
+      if (!CommonChecks.isValidID(nodeId)) {
+        result.addError("Invalid node id: " + nodeId);
+      }
+    }
+
+    // Check for duplicate node IDs (this is already guaranteed by using HashMap, but good to be
+    // explicit)
+    // No need for additional check since HashMap ensures uniqueness
+
+    // Ensuring that containments and annotations are the inverse of parent relationships
+    Map<String, Set<String>> containedNodes = new HashMap<>();
+    for (SerializedClassifierInstance node : nodesByID.values()) {
+      for (SerializedContainmentValue containmentValue : node.getContainments()) {
+        for (String childId : containmentValue.getChildrenIds()) {
+          // Verifying nodes do not appear in multiple containments or annotations
+          String newPlacement = node.getID() + " at " + containmentValue.getMetaPointer();
+          if (containedNodes.containsKey(childId)) {
+            result.addError(
+                childId
+                    + " is listed in multiple places: "
+                    + containedNodes.get(childId)
+                    + " and now "
+                    + newPlacement);
+          } else {
+            containedNodes.put(childId, new HashSet<>(Arrays.asList(newPlacement)));
+          }
+          SerializedClassifierInstance child = nodesByID.get(childId);
+          if (child != null && !child.getParentNodeID().equals(node.getID())) {
+            result.addError(
+                childId
+                    + " is listed as child of "
+                    + node.getID()
+                    + " but it has as parent "
+                    + child.getParentNodeID());
+          }
+        }
+      }
+      for (String annotationId : node.getAnnotations()) {
+        // Verifying nodes do not appear in multiple containments or annotations
+        String newPlacement = node.getID() + " among annotations";
+        if (containedNodes.containsKey(annotationId)) {
+          result.addError(
+              annotationId
+                  + " is listed in multiple places: "
+                  + containedNodes.get(annotationId)
+                  + " and now "
+                  + newPlacement);
+        } else {
+          containedNodes.put(annotationId, new HashSet<>(Arrays.asList(newPlacement)));
+        }
+        SerializedClassifierInstance annotation = nodesByID.get(annotationId);
+        if (annotationId != null
+            && annotation != null
+            && !Objects.equals(annotation.getParentNodeID(), node.getID())) {
+          result.addError(
+              annotationId
+                  + " is listed as an annotation of "
+                  + node.getID()
+                  + " but it has as parent "
```

Comment:
Can just as well phrase it as `" but it has ${parentID} as parent"`.

Reply:
Changes as suggested